// @testing-library/react-hooks/lib/helpers/createTestHarness.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createTestHarness = createTestHarness;

var _react = _interopRequireWildcard(require("react"));

var _reactErrorBoundary = require("react-error-boundary");

var _filterConsole = _interopRequireDefault(require("filter-console"));

var _core = require("../core");

function suppressErrorOutput() {
  // The error output from error boundaries is notoriously difficult to suppress.  To save
  // out users from having to work it out, we crudely suppress the output matching the patterns
  // below.  For more information, see these issues:
  //   - https://github.com/testing-library/react-hooks-testing-library/issues/50
  //   - https://github.com/facebook/react/issues/11098#issuecomment-412682721
  //   - https://github.com/facebook/react/issues/15520
  //   - https://github.com/facebook/react/issues/18841
  const removeConsoleFilter = (0, _filterConsole.default)([/^The above error occurred in the <TestComponent> component:/, // error boundary output
  /^Error: Uncaught .+/ // jsdom output
  ], {
    methods: ['error']
  });
  (0, _core.addCleanup)(removeConsoleFilter);
}

function createTestHarness({
  callback,
  setValue,
  setError
}, Wrapper, suspense = true) {
  const TestComponent = ({
    hookProps
  }) => {
    // coerce undefined into TProps, so it maintains the previous behaviour
    setValue(callback(hookProps));
    return null;
  };

  let resetErrorBoundary = () => {};

  const ErrorFallback = ({
    error,
    resetErrorBoundary: reset
  }) => {
    resetErrorBoundary = () => {
      resetErrorBoundary = () => {};

      reset();
    };

    setError(error);
    return null;
  };

  suppressErrorOutput();

  const testHarness = props => {
    resetErrorBoundary();

    let component = /*#__PURE__*/_react.default.createElement(TestComponent, {
      hookProps: props
    });

    if (Wrapper) {
      component = /*#__PURE__*/_react.default.createElement(Wrapper, props, component);
    }

    if (suspense) {
      component = /*#__PURE__*/_react.default.createElement(_react.Suspense, {
        fallback: null
      }, component);
    }

    return /*#__PURE__*/_react.default.createElement(_reactErrorBoundary.ErrorBoundary, {
      FallbackComponent: ErrorFallback
    }, component);
  };

  return testHarness;
}
